### ISSUE TITLE
Crash with core dump

### ISSUE THREAD

**Issue by npyoung** [ray-project/ray#3078]
I'm getting a crash when I scale up from small test/example code to my actual task on ray. It's not clear to me from the error messages and logs what happened and why.

### System information
- **OS Platform and Distribution**: Ubuntu 16.04 (EC2 Deep Learning AMI)
- **Ray installed from**: binary
- **Ray version**: 0.5.3
- **Python version**: 3.6
- Using raylet (`ray start --use-raylet` set, and `ray.init(..., use_raylet=True)` in script)

Broadly, I run about 10 tasks with `num_gpus=0.25` which each return a large numpy array, or rather a ray object ID, since I don't pull these arrays back to the driver code. Then I pass a list of these IDs, along with some unique parameters, into a large number of ray tasks with `num_gpus=0.5`. Each of these secondary tasks read all the numpy arrays from the first batch, but don't write or copy them.

I've tested GPU tasks, fractional GPU usage, and ray's ability to trace object dependencies on my cluster and it seems like in small examples all these things are fine. When I run my real code, I get several

```<id> has been marked dead because the monitor has missed too many heartbeats```

followed by
```
WARNING: Logging before InitGoogleLogging() is written to STDERR
F1017 20:46:36.214735 14867 local_scheduler_client.cc:321]  Check failed: static_cast<ray::protocol::MessageType>(type) == ray::protocol::MessageType::WaitReply
*** Check failure stack trace: ***
Aborted (core dumped)
```

If I try to restart the task (not ray), I see
```
WARNING: Logging before InitGoogleLogging() is written to STDERR
E1017 20:55:28.937588 17097 io.cc:119] Connection to socket failed for pathname /tmp/ray/session_2018-10-17_19-41-14_2598/sockets/raylet
F1017 20:55:33.941112 17097 io.cc:127] Could not connect to socket /tmp/ray/session_2018-10-17_19-41-14_2598/sockets/raylet
*** Check failure stack trace: ***
Aborted (core dumped)
```

### Logs
- Worker logs are empty
- Raylet logs give lots of `Failed to forward task` notices
- Plasma logs say `store to use up to 13.3GB of memory`. I should need about 9GB.
- Monitor logs say that resource usage while task is running is `0.0/20.0 CPU, 2.0/2.0 GPU` with `2/2 target nodes` and 2 nodes used. (Cluster should have 2-8 workers.)

**Comment by richardliaw** [comment]
Hm, any way we can reproduce this?

**Comment by npyoung** [comment]
I was able to reliably reproduce this by putting big enough arrays into the object store. I can also reliably solve this problem by switching to instances with more memory. It looked like the plasma store is large enough, but I am now convinced this is an out-of-memory issue. I don't think system memory is filling up, but I do think shared memory might fill up.

It seems like running out of memory a) stops not just the task, but crashes ray and b) doesn't get reported in any of the logs. Is it possible to make ray survive this situation and report a useful error somewhere?

Question: is the object store mirrored across all nodes or do machines only get data as tasks request it? In my case I have some fairly large arrays, but they are created and used by GPU tasks, so I assumed the head node could get away with less system memory. I had to add memory to both workers and head to get my code to run.
